src/python/pants/jvm/resolve/coursier_fetch.py

```diff
@@ -0,0 +1,526 @@
+# Copyright 2021 Pants project contributors (see CONTRIBUTORS.md).
+# Licensed under the Apache License, Version 2.0 (see LICENSE).
+
+from __future__ import annotations
+
+import json
+import logging
+import os
+from dataclasses import dataclass
+from pathlib import PurePath
+from typing import Iterable, Optional, Tuple
+
+from pants.core.util_rules.source_files import SourceFiles, SourceFilesRequest
+from pants.engine.collection import Collection, DeduplicatedCollection
+from pants.engine.fs import (
+    AddPrefix,
+    Digest,
+    DigestContents,
+    DigestSubset,
+    FileDigest,
+    MergeDigests,
+    PathGlobs,
+    RemovePrefix,
+)
+from pants.engine.process import Process, ProcessResult
+from pants.engine.rules import Get, MultiGet, collect_rules, rule
+from pants.engine.target import Targets, TransitiveTargets, TransitiveTargetsRequest
+from pants.jvm.resolve.coursier_setup import Coursier
+from pants.jvm.target_types import JvmLockfileSources, MavenRequirementsField
+from pants.jvm.util_rules import ExtractFileDigest
+from pants.util.logging import LogLevel
+
+logger = logging.getLogger(__name__)
+
+
+class CoursierError(Exception):
+    """An exception relating to invoking Coursier or processing its output."""
+
+
+class MavenRequirements(DeduplicatedCollection[str]):
+    @classmethod
+    def create_from_maven_coordinates_fields(
+        cls,
+        fields: Iterable[MavenRequirementsField],
+        *,
+        additional_requirements: Iterable[str] = (),
+    ) -> MavenRequirements:
+        field_requirements = (
+            str(maven_coord) for field in fields for maven_coord in (field.value or ())
+        )
+        return MavenRequirements((*field_requirements, *additional_requirements))
+
+
+@dataclass(frozen=True)
+class MavenCoord:
+    """A single Maven-style coordinate for a JVM dependency."""
+
+    # TODO: parse and validate the input into individual coordinate
+    # components, then re-expose the string coordinate as a property
+    # or __str__.
+    coord: str
+
+
+class MavenCoordinates(Collection[MavenCoord]):
```

Comment:
FYI you can also use `DeduplicatedCollection` to newtype `FrozenOrderedSet`, giving you deduplication. You can set `sort_input = True` too so that inputs to other rules are more deterministic.